obspy/clients/filesystem/tsindex.py

```diff
@@ -0,0 +1,1743 @@
+                           .format(self.tsindex_summary_table))
+
+        connection.execute(
+            "CREATE {0} TABLE {1} AS "
+            "SELECT network, station, location, channel, "
+            "  min(starttime) AS earliest, max(endtime) AS latest, "
+            "  datetime('now') as updt "
+            "FROM {2} "
+            "GROUP BY 1,2,3,4;"
+            .format("TEMPORARY" if temporary is True else "",
+                    self.tsindex_summary_table,
+                    self.tsindex_table)
+        )
+        if connection is None:
+            connection.close()
+
+    def has_tsindex_summary(self, connection=None):
+        """
+        Returns True if there is a tsindex_summary table in the database.
+
+        :type connection: sqlalchemy.engine.Connection
+        :param: connection: SQLAlchemy database connection object.
+        :rtype: bool
+        :returns: Returns True if there a tsindex_summary table is present
+            in the database.
+        """
+        if connection is None:
+            connection = self.engine.connect()
+        result = connection.execute("SELECT count(*) FROM sqlite_master "
+                                    "WHERE type='table' and name='{0}'"
+                                    .format(self.tsindex_summary_table))
+
+        summary_present = result.fetchone()[0]
+        if connection is None:
+            connection.close()
+        if summary_present:
+            return True
+        else:
+            return False
+
+    def has_tsindex(self, connection=None):
+        """
+        Returns True if there is a tsindex table in the database.
+
+        :type connection: sqlalchemy.engine.Connection
+        :param: connection: SQLAlchemy database connection object.
+        :rtype: bool
+        :returns: Returns True if there a tsindex_summary table is present
+            in the database.
+        """
+        if connection is None:
+            connection = self.engine.connect()
+        result = connection.execute("SELECT count(*) FROM sqlite_master "
+                                    "WHERE type='table' and name='{0}'"
+                                    .format(self.tsindex_table))
+
+        summary_present = result.fetchone()[0]
+        if connection is None:
+            connection.close()
+        if summary_present:
+            return True
+        else:
+            return False
+
+    def _open_database_connection(self):
+        try:
+            connection = self.engine.connect()
+        except Exception as err:
+            raise ValueError(str(err))
+
+        logger.debug("Opening SQLite database for "
+                     "index rows: %s" % self.sqlitedb)
+
+        # Store temporary table(s) in memory
+        try:
+            connection.execute("PRAGMA temp_store=MEMORY")
+        except Exception as err:
+            raise ValueError(str(err))
+        return connection
+
+    def _fetch_index_rows(self, query_rows=[], bulk_params={}):
```

Comment:
I'd strongly advise against mutables as kwarg default values. It's too easy to step into traps and end up in time consuming debugging. Better use `None` and set empty list (or dict etc.) in an `if .. is None:` check.